rclpy/src/rclpy/_rclpy.c

```diff
@@ -879,6 +880,99 @@ rclpy_count_subscribers(PyObject * Py_UNUSED(self), PyObject * args)
   return _count_subscribers_publishers(args, "subscribers", rcl_count_subscribers);
 }
 
+typedef rcl_ret_t (* rcl_get_info_by_topic_func_t)(
+  const rcl_node_t * node,
+  rcutils_allocator_t * allocator,
+  const char * topic_name,
+  bool no_mangle,
+  rmw_topic_endpoint_info_array_t * info_array);
+
+static PyObject *
+_get_info_by_topic(
+  PyObject * args, const char * type,
+  rcl_get_info_by_topic_func_t rcl_get_info_by_topic)
+{
+  PyObject * pynode;
+  const char * topic_name;
+  int no_mangle;
+
+  if (!PyArg_ParseTuple(args, "Osp", &pynode, &topic_name, &no_mangle)) {
+    return NULL;
+  }
+
+  rcl_node_t * node = (rcl_node_t *)PyCapsule_GetPointer(pynode, "rcl_node_t");
+  if (!node) {
+    return NULL;
+  }
+  rcutils_allocator_t allocator = rcutils_get_default_allocator();
+  rmw_topic_endpoint_info_array_t info_array = rmw_get_zero_initialized_topic_endpoint_info_array();
+  rcl_ret_t ret = rcl_get_info_by_topic(node, &allocator, topic_name, no_mangle, &info_array);
+  rmw_ret_t fini_ret;
+  if (ret != RCL_RET_OK) {
+    if (ret == RCL_RET_BAD_ALLOC) {
+      PyErr_Format(PyExc_MemoryError, "Failed to get information by topic for %s: %s",
+        type, rcl_get_error_string().str);
+    } else if (ret == RCL_RET_UNSUPPORTED) {
+      PyErr_Format(PyExc_NotImplementedError, "Failed to get information by topic for %s: "
+        "function not supported by RMW_IMPLEMENTATION", type);
+    } else {
+      PyErr_Format(RCLError, "Failed to get information by topic for %s: %s",
+        type, rcl_get_error_string().str);
+    }
+    rcl_reset_error();
+    fini_ret = rmw_topic_endpoint_info_array_fini(&info_array, &allocator);
```

Comment:
Is there not an rcl equivalent to this function? It's weird to be mixing rcl and rmw, e.g. `rcl_get_info_by_topic()` and `rmw_topic_endpoint_info_array_fini()`.

Often we just typedef or `#define` the rmw types and functions to be rcl ones, that way we could extend just the rcl layer if we want to do so with minimal changes to users of rcl (like rclpy and rclcpp).

As much as possible, rclpy and rclcpp should avoid using rmw directly and depend on rcl only, even if the items in rcl are just typedefs to rmw. This is one of the things that is missed when you split merging of rmw/rcl changes from rclpy/rclcpp and higher level demo changes...

Reply:
Changed everything to using `rcl` types.

I don't think this is too big of a problem caused by splitting the merging of `rmw`/`rcl` and `rclpy`/`rclcpp`, because it will only require a small and purely additive change to `rcl` afterwards (https://github.com/ros2/rcl/pull/558).